src/Pos/Block/Logic.hs

```diff
@@ -883,16 +887,32 @@ createMainBlockPure
     -> m (MainBlock ssc)
 createMainBlockPure limit prevHeader txs pSk sId psks sscData usPayload sk =
     flip evalStateT limit $ do
-        -- account for block header and serialization overhead, etc; also
-        -- include all SSC data because a) deciding is hard and b) we don't
-        -- yet have a way to strip generic SSC data
+        -- default ssc to put in case we won't fit a normal one
+        let defSsc = (sscDefaultPayload @ssc (siSlot sId) :: SscPayload ssc)
+
+        -- account for block header and serialization overhead, etc;
         let musthaveBody = Types.MainBody
                 (fromMaybe (error "createMainBlockPure: impossible") $ mkTxPayload mempty)
-                sscData [] def
+                defSsc [] def
         musthaveBlock <-
             either throwError pure $
             mkMainBlock (Just prevHeader) sId sk pSk musthaveBody extraH extraB
-        count musthaveBlock
+        let mhbSize = biSize musthaveBlock
+        when (mhbSize > fromIntegral limit) $ throwError $
```

Comment:
Btw, `fromIntegral` shouldn't be necessary here and in some other places anymore.